Tweak some tests under my point of view.

Guardiola31337 opened the pull request:
I tweaked some tests using DAMP (Descriptive And Maintainable Procedures) because I prefer use these techniques over DRY (Don't Repeat Yourself). In my opinion, the way you should treat test code is not the same as the way we are accustomed to write production code. So, I prefer this way because using DAMP you increase maintainability and readability of the tests.

Commit:
tweak some tests under my point of view

pedrovgs commented:
Review this PR please, you have checkstyle errors :)

Commit:
fix checkstyle errors

Guardiola31337 commented:
Done!

pedrovgs commented:
Hi Pablo! 

First of all, thanks for your time and this PR :)

I've been reviewing your PR and the usage of DAMP over DRY seems interesting. There are some things I'd like to discuss with you before to merge this PR.

I understand the test engineering has different rules and patterns we don't apply to write production code but I see some points where the code is not so good as it should be. The key of the following points to discuss is the one @cookingsource told me few seconds ago discussing about this PR: "why do you assume DRY principle is going to be used to write less readable and less descriptive code?"

**1. setUp() method.** You've removed the setUp method and moved that code to the test arrange zone. This is a sample:

``` java

HeaderRecyclerViewAdapter<VH, H, T> adapterWithHeaderAndWithFiveItems = getAdapterUnderTest();
List<T> items = givenAListWithFiveItems();
H header = givenAHeader();
adapterWithHeaderAndWithFiveItems.setHeader(header);
adapterWithHeaderAndWithFiveItems.setItems(items);

```

Instead of duplicate all this code why don't you move this code into a private method `givenAnAdapterWithHeaderAndFiveItems()` or leave the adapter initialization into the `setUp()` method? This code is a little bit more readable, but you are duplicating code. Using a method you could have the good parts of both approaches, don't repeat code and at the same team improve the test readability. If you don't want to write one method per test set up you can create a `givenAnAdapter` method with different params or a builder. In fact, you are already doing it in some of the last tests like `shouldDelegateCallToOnBindItemViewHolderIfViewTypeIsItemType`

**2. Expecting literals.** You've removed all the constants in this file used as expected values. In this case, expect a 1 or a 0 it's ok because that namers has semantic but with other values like the type associated to a position into the adapter `TYPE_HEADER` or `TYPE_ITEM` I don't understand why you've removed this constants to expect -1 or -2. That constants were declared into the test and not in the SUT, to me you are not adding readability with this change. Here you have a sample:

``` java

 assertEquals(-1, adapterWithoutHeaderAndWithoutItems.getItemViewType(anyPosition));

```

What does it mean -1 there? Don't you prefer this approach?

``` java

 assertEquals(ITEM_TYPE, adapterWithoutHeaderAndWithoutItems.getItemViewType(anyPosition));

```

**3. Variable names.** You've removed the **adapter** class member, the SUT in this test cases and created different adapter variables with different names like: `adapterWithoutHeaderAndWithoutItems`. To me this information is already implicit in the test context because the name is descriptive enough. When you are reading that test you already have in mind what you are going to test just reading the test name and provide more information using a long variable name is not improving the test readability. Here you have a sample:

``` java

 @Test
    public void shouldDelegateCallToOnBindItemViewHolderIfViewTypeIsItemType() {
        HeaderRecyclerViewAdapter<VH, H, T> adapterWithHeaderAndSomeItems = spy(givenAnAdapterWithHeaderAndSomeItems());
        VH holder = givenAViewHolder();
        int anyNonHeaderPosition = 1;

        adapterWithHeaderAndSomeItems.onBindViewHolder(holder, anyNonHeaderPosition);

        verify(adapterWithHeaderAndSomeItems).onBindItemViewHolder(holder, anyNonHeaderPosition);
}

```

Don't you prefer to replace `adapterWithHeaderAndSomeItems` with just `adapter` as variable name? The name of this test: `shouldDelegateCallToOnBindItemViewHolderIfViewTypeIsItemType` and the given method:  `givenAnAdapterWithHeaderAndSomeItems` already provides the context information you need to understand the test.

**4. Using variables instead of constants.** In this PR I can see some variables like: `int positionZero = 0` don't you prefer to use just the value `0` descriptive enough when you work with lists or a constant like `POSITION_ZERO` or `FIRST_POSITION`? To me, duplicate variables in this tests just to replace the literal with a variable instead of use the value (in this concrete case) or a constant is not going to improve the code readability.

Thanks for this PR :) It's really funny tu discuss abut testing with a real sample.

Cookingsource commented:
_Moving  here from a commit comment (Related to long test names )_
I think this kind of naming is a code smell. In this case it's hiding the fact that you have several families of scenarios to test, and they are orthogonal.

I understand that in some cases setting up the preconditions for a test is a complex task, sometimes unavoidable (for example the input structure is indeed a complex structure by definition ie. A graph ) In those cases you can take advantage of inheritance to setup common preconditions for a familly of tests. This is good because of several reasons.
- It's easier to refactor the way to generate that subscenario preconditions
- The tests have shorter names
- All tests of a subscenario remain together

One could argue that this class is small to deal with this, but you are almost already doing it by making the contract tests as a base class. You could deepen the abstract hierarchy and take advantage of it in the chained setups. They only thing to have in mind if you don't want to repeat the run of tests is to put only the actual test on leafs of the hierarchy ( in your case because of the contracts, it's ok to be one depth far from the leaf)

Let me put an example for testing a pathfinder. It's input is a graph. The tests are structured this way
- PathFinderTests
- PathFinderTests<-DegenerateCases
- PathFinderTests<-SingleDirectionEdge
- PathFinderTests<-SequentialGraph

In this hierarchy you could take advantage of the setup to configure common preconditions for each subscenario

Commit:
add HeaderRecyclerViewAdapter builder

Commit:
refactor tests using HeaderRecyclerViewAdapterBuilder

Commit:
remove methods never used and fix formatting

Commit:
add GridLayoutManager builder

Commit:
add HeaderSpanSizeLookup builder

Commit:
refactor tests using builders

Commit:
remove method never used

Commit:
extract TYPE_ITEM and TYPE_HEADER constants and left it as it was

Commit:
fix variable names from adapterWithHeaderAndSomeItems to adapter when necessary

Commit:
extract constants POSITION_ZERO, ANY_POSITION, POSITION_ONE and HEADER_POSITION

Commit:
extract constants THREE_ROWS, POSITION_ZERO and POSITION_ONE

Guardiola31337 commented:
First of all, as I pointed out in the commit message, what I modified was from my point of view.
I wanted to show my perspective, that's all. I didn't want it to be applied blindly. Perhaps it was my fault because I exaggerate it in some parts.

That said, I know there is not much literature on the subject so I agree with you that it will be funny to discuss around this sample ;)

Most of my thoughts are based on the awesome book [Working Effectively With Unit Tests](http://wewut.com/). I really like the techniques described by @jaycfields on it in favor of DAMP.

So let me explain each point:

**1. setUp() method**
- When you remove Setup you may limit the number of times you need to check variables when you first encounter a test. So then you no longer need to worry about irrelevant instances from other tests
- It's a waste of time because you normally need to understand all the Setup each time you have to deal with the test
- Setup advocates argue because if you don't use it you normally create duplication. As you mentioned you can solve it using private methods or builders, I personally prefer builders so I did [some refactors](https://github.com/Guardiola31337/HeaderRecyclerView/pull/1) in order to improve our test suite

**2. Expecting literals**
- The main advantage of expecting literals is that you only need to focus on the actual value of the assertion, increasing readability and traceability. 
- In this particular case it could be more explanatory using the constants you proposed in the beginning so I left it as it was :wink:

**3. Variable names**
- I agree in some, but not all, the cases.
- When the variable is a collaborator needed in the context defined but it doesn't provide you any extra semantics in the test I would use `adapter`
  
  ``` java
  @Test public void shouldDelegateCallToOnCreateHeaderViewHolderIfViewTypeIsHeaderType() {
    HeaderRecyclerViewAdapter<VH, H, T> adapter =
        spy(new HeaderRecyclerViewAdapterBuilder().withHeader(givenAHeader())
            .withItems(givenAListWithFiveItems())
            .build());
    ViewGroup anyViewGroup = mock(ViewGroup.class);
  
    adapter.onCreateViewHolder(anyViewGroup, TYPE_HEADER);
  
    verify(adapter).onCreateHeaderViewHolder(anyViewGroup, TYPE_HEADER);
  }
  ```
  
  In the above test what we are testing has nothing to do with how the `adapter` is configured, so I modified those cases
  
  But, if we analyze the following test:
  
  ``` java
  @Test public void shouldDelegateCallToOnBindHeaderViewHolderIfViewTypeIsHeaderType() {
    HeaderRecyclerViewAdapter<VH, H, T> adapterWithHeaderAndSomeItems =
        spy(new HeaderRecyclerViewAdapterBuilder().withHeader(givenAHeader())
            .withItems(givenAListWithFiveItems())
            .build());
    VH holder = givenAViewHolder();
    int headerPosition = 0;
  
    adapterWithHeaderAndSomeItems.onBindViewHolder(holder, headerPosition);
  
    verify(adapterWithHeaderAndSomeItems).onBindHeaderViewHolder(holder, headerPosition);
  }
  ```
  
  `onBindViewHolder` method depends on how the `adapter` is configured, in fact, if it has a header or not.
  So in cases like these I prefer to describe the config through naming

**4. Using variables instead of constants**
- I completely agree with you. Instead of using explaining variables I should have used constants
- This minor refactor wasn't related with the tests themselves, so it was out of my scope. Anyway, I fixed it as you suggested :thumbsup:

I'm glad to discuss about these tiny things. I consider them really important to learn and grow into better programmers.
I would like to remind everyone that if you reach this level of granularity you are doing it really really well.
But it's good to talk about them and strive to improve our craft.
It would be a pleasure to go into more depth drinking some :beer::beer:
I'm a strong believer in collaborative learning so please feel free to join the discussion and share with us your opinion.

pedrovgs commented on library/src/test/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapterBaseTest.java:
```diff
@@ -37,145 +36,142 @@
+  public static final int ANY_POSITION = 3;
+  public static final int POSITION_ONE = 1;
+  public static final int HEADER_POSITION = 0;
+  public static final int ANY_NON_HEADER_POSITION = 4;
```
This constants should be private.

Guardiola31337 replied:
You are right. I did it within the Extract>Constant of the IDE and the IDE extracts them as public.

pedrovgs commented on library/src/test/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapterBaseTest.java:
```diff
@@ -37,145 +36,142 @@
 
-    int expectedSize = items.size() + 1;
-    Assert.assertEquals(expectedSize, adapter.getItemCount());
+    Assert.assertEquals(6, adapterWithHeaderAndWithFiveItems.getItemCount());
```
Do you need this Assert class here? Can't you use just assertEquals(6......)?

Guardiola31337 replied:
No, I don't. I didn't modify that at all. I asked myself the same question.
So, sure, I change it.

pedrovgs commented:
This PR is almost done. Take a look to the last two comments I've added and we will merge it :)

@flipper83 do you have something to add in this PR?

Commit:
fix PR comments

Commit:
fix checkstyle errors

Guardiola31337 commented:
Anytime! :satisfied:

pedrovgs commented:
:+1:  from my side. We are going to wait for @flipper83 to merge this PR. Thanks @Guardiola31337 !

flipper83 commented:
:+1:
